Amplitude/AMPDatabaseHelper.m

```diff
@@ -0,0 +1,502 @@
+        success &= [db executeUpdate: dropLongStoreTableSQL];
+
+        [db close];
+    }];
+
+    return success;
+}
+
+- (BOOL)resetDB:(BOOL) deleteDB
+{
+    BOOL success = YES;
+
+    if (deleteDB) {
+        success &= [self deleteDB];
+    } else {
+        success &= [self dropTables];
+    }
+    success &= [self createTables];
+
+    return success;
+}
+
+- (BOOL)deleteDB
+{
+    BOOL success = YES;
+    if ([[NSFileManager defaultManager] fileExistsAtPath:_databasePath] == YES) {
+        success = [[NSFileManager defaultManager] removeItemAtPath:_databasePath error:NULL];
+    }
+    return success;
+}
+
+- (BOOL)addEvent:(NSString*) event
+{
+    return [self addEventToTable:EVENT_TABLE_NAME event:event];
+}
+
+- (BOOL)addEventToTable:(NSString*) table event:(NSString*) event
+{
+    __block BOOL success = NO;
+    __block NSString *errMsg;
+
+    [_dbQueue inDatabase:^(FMDatabase *db) {
+        if (![db open]) {
+            NSLog(@"Failed to open database during addEventToTable %@", table);
+            return;
+        }
+
+        NSString *insertSQL = [NSString stringWithFormat:INSERT_EVENT, table, EVENT_FIELD];
+        success = [db executeUpdate:insertSQL, event];
+        if (!success) {
+            errMsg = [db lastErrorMessage];
+        }
+        [db close];
+    }];
+
+    if (!success) {
+        NSLog(@"addEventToTable %@ failed: %@", table, errMsg);
+        [self resetDB:NO]; // not much we can do, just start fresh
+    }
+    return success;
+}
+
+- (NSDictionary*)getEvents:(long) upToId limit:(long) limit
+{
+    return [self getEventsFromTable:EVENT_TABLE_NAME upToId:upToId limit:limit];
+}
+
+- (NSDictionary*)getEventsFromTable:(NSString*) table upToId:(long) upToId limit:(long) limit
+{
+    __block long maxId = -1;
+    __block NSMutableArray *events = [NSMutableArray array];
+
+    [_dbQueue inDatabase:^(FMDatabase *db) {
+        if (![db open]) {
+            NSLog(@"Failed to open database during getEventsFromTable %@", table);
+            return;
+        }
+
+        NSString *querySQL;
+        if (upToId > 0 && limit > 0) {
+            querySQL = [NSString stringWithFormat:GET_EVENT_WITH_UPTOID_AND_LIMIT, ID_FIELD, EVENT_FIELD, table, ID_FIELD, upToId, limit];
+        } else if (upToId > 0) {
+            querySQL = [NSString stringWithFormat:GET_EVENT_WITH_UPTOID, ID_FIELD, EVENT_FIELD, table, ID_FIELD, upToId];
+        } else if (limit > 0) {
+            querySQL = [NSString stringWithFormat:GET_EVENT_WITH_LIMIT, ID_FIELD, EVENT_FIELD, table, limit];
+        } else {
+            querySQL = [NSString stringWithFormat:GET_EVENT, ID_FIELD, EVENT_FIELD, table];
+        }
+        FMResultSet *s = [db executeQuery:querySQL];
+        if (s == nil) {
+            NSLog(@"getEvents from table %@ failed: %@", table, [db lastErrorMessage]);
+            [db close];
+            [self resetDB:NO];
+            return;
+        }
+
+        while ([s next]) {
```

Comment:
How cool is this, closing the DB closes all open resultsets: https://github.com/ccgus/fmdb/commit/ae8622d5b05747bc4573b1f0520b1b346a94575e